File: lib/bloc/user/UserBloc.java
class UserBloc extends Bloc<UserEvent, UserState> {
  UserBloc({required this.userRepository}) : super(UserState.initial());

  final UserRepository userRepository;

  @override
  Stream<UserState> mapEventToState(UserEvent event) async* {
    if (event is Login) {
      yield UserLoading(user: state.user);
      try {
        final user = await userRepository.login(event.username, event.password);
        yield UserLoaded(user: user);
      } catch (e) {
        yield UserLoadFailure(e);
      }
    } else if (event is Logout) {
      yield UserLoading(user: state.user);
      try {
        await userRepository.logout();
        yield UserState.initial();
      } catch (e) {
        yield UserLoadFailure(e);
      }
    }
  }
}

class UserEvent extends Equatable {
  const UserEvent();

  @override
  List<Object> get props => [];
}

class Login extends UserEvent {
  const Login({required this.username, required this.password});

  final String username;
  final String password;

  @override
  List<Object> get props => [username, password];
}

class Logout extends UserEvent {
  const Logout();
}

File: lib/bloc/user/user_bloc.dart
import 'dart:async';

import 'package:Array_App/bloc/user/user_event.dart';
import 'package:Array_App/bloc/user/user_state.dart';
import 'package:bloc/bloc.dart';
import 'package:stream_transform/stream_transform.dart';

import '../../domain/use_case/initial_item_load_use_case.dart';

// class InitialLoadUseCase {
//   final UserRepository userRepository;
//   final ItemRepository itemRepository;
//
//   InitialLoadUseCase({this.userRepository, this.itemRepository});
//
//   Future<Data> loadData() async {
//     final user = await userRepository.getUser();
//     final items = await itemRepository.getItems();
//     return Data(user: user, items: items);
//   }
// }
// TODO(jtl) probably in initial page the user will load and
// then the date will initiate a itemloadevent.
class UserBloc extends Bloc<UserEvent, UserState> {
  UserBloc(
    this._initialLoadUseCase,
  ) : super(const UserState.initial()) {
    on<LoadUser>(
      _onLoadStarted,
      transformer: (events, mapper) => events.switchMap(mapper),
    );
  }

  final InitialItemLoadUseCase _initialLoadUseCase;

  Future<void> _onLoadStarted(UserEvent event, Emitter<UserState> emit) async {
    try {
      emit(UserLoading(user: state.user));
      // TODO(jtl): Implement this.
      // final user = await _initialLoadUseCase.execute(1);
      // emit(UserLoaded(user: user));
    } on Exception catch (e) {
      emit(UserLoadFailure(e));
    }
  }
}
